src/main/java/com/exactpro/th2/codec/qfj/QFJCodec.java

```diff
@@ -0,0 +1,473 @@
+                        .setId(msgMetadata.getId())
+                        .setTimestamp(msgMetadata.getTimestamp())
+                        .build())
+                .build();
+    }
+
+    private quickfix.Message getFixMessage(Map<String, Value> fieldsMap, String msgName) {
+
+        boolean isFix50 = false;
+        if (appDataDictionary != null) {
+            isFix50 = true;
+            localDataDictionary = appDataDictionary;
+        } else if (dataDictionary != null) {
+            localDataDictionary = dataDictionary;
+        } else {
+            throw new IllegalStateException("No available dictionaries.");
+        }
+
+        String msgType = localDataDictionary.getMsgType(msgName);
+
+        quickfix.Message message = new quickfix.Message();
+        quickfix.Message.Header header = message.getHeader();
+        quickfix.Message.Trailer trailer = message.getTrailer();
+
+        if (!fieldsMap.containsKey("Header")) {
+            header.setField(new BeginString(localDataDictionary.getVersion()));
+        }
+
+        String key;
+        int tag;
+        Value value;
+
+        for (Map.Entry<String, Value> item : fieldsMap.entrySet()) {
+
+            key = item.getKey();
+            value = item.getValue();
+
+            if (key.equals("Header")) {
+                setHeader(value, header, isFix50);
+            } else if (key.equals("Trailer")) {
+                setTrailer(value, trailer, isFix50);
+            } else {
+
+                tag = validateTag(localDataDictionary.getFieldTag(key), key);
+
+                if (isGroup(value)) {
+                    DataDictionary groupDataDictionary = localDataDictionary.getGroup(msgType, tag).getDataDictionary();
+                    for (Group group : getGroups(value, tag, localDataDictionary, groupDataDictionary, msgType)) {
+                        message.addGroup(group);
+                    }
+                } else {
+                    if (isHeaderField(tag, isFix50)) {
+                        throw new IllegalArgumentException("Header tag \"" + key + "\" in the body");
+                    } else if (isTrailerField(tag, isFix50)) {
+                        throw new IllegalArgumentException("Trailer tag \"" + key + "\" in the body");
+                    } else {
+                        if (!localDataDictionary.isMsgField(msgType, tag)) {
+                            throw new IllegalArgumentException("Tag \"" + key + "\" does not belong to this type of message: " + msgType);
+                        }
+                        Field<?> field = new Field<>(tag, value.getSimpleValue());
+                        message.setField(tag, field);
+                    }
+                }
+            }
+        }
+        header.setField(new MsgType(msgType));
+
+        return message;
+    }
+
+    private List<Group> getGroups(Value value, int tag, DataDictionary dataDictionary, DataDictionary groupDictionary, String msgType) {
```

Comment:
```suggestion
    private List<Group> getGroups(List<Value> values, int tag, DataDictionary dataDictionary, DataDictionary groupDictionary, String msgType) {
```
> DataDictionary dataDictionary

Why do we need it? I thought that group dictionary should contain info about all group fields